### ISSUE TITLE
[Windows] Create named pipe for docker client to connect

### ISSUE THREAD

**Issue by praveenkumar** [code-ready/crc#3018]
https://docs.docker.com/desktop/faqs/#how-do-i-connect-to-the-remote-docker-engine-api in windows docker client use named pipe to connect docker API and unix socket doesn't work by default. We need to create a proxy which able to proxy the named pipe to unix socket.  I am also not able to find if using ssh tunnel we can directly forward a named pipe one end and docker sock on other end (may be not supported at all). 

In gvisor-tap-vsock code have implementation around forwarding npipe over ssh but not around tunneling, need to dig bit more around it. 

- https://github.com/containers/gvisor-tap-vsock/blob/main/pkg/sshclient/ssh_forwarder.go#L131-L180

**Comment by cfergeau** [comment]
https://github.com/containers/gvisor-tap-vsock/blob/main/pkg/services/forwarder/ports.go would need to be modified to add support for some magic URL to do npipe<->unix tunneling/forwarding.

**Comment by cfergeau** [comment]
I wonder if this could be relatively simple:
```
diff --git a/pkg/services/forwarder/ports.go b/pkg/services/forwarder/ports.go
index 36c300f..43de6f9 100644
--- a/pkg/services/forwarder/ports.go
+++ b/pkg/services/forwarder/ports.go
@@ -9,7 +9,6 @@ import (
 	"net"
 	"net/http"
 	"net/url"
-	"os"
 	"sort"
 	"strconv"
 	"strings"
@@ -73,7 +72,7 @@ func (f *PortsForwarder) Expose(protocol types.TransportProtocol, local, remote
 	}
 
 	switch protocol {
-	case types.UNIX:
+	case types.NPIPE:
 		// parse URI for remote
 		remoteURI, err := url.Parse(remote)
 		if err != nil {
@@ -155,13 +154,13 @@ func (f *PortsForwarder) Expose(protocol types.TransportProtocol, local, remote
 
 		// build the tcp proxy
 		var p tcpproxy.Proxy
-		p.ListenFunc = func(_, socketPath string) (net.Listener, error) {
-			// remove existing socket file
-			if err := os.Remove(socketPath); err != nil && !os.IsNotExist(err) {
+		p.ListenFunc = func(_, npipePath string) (net.Listener, error) {
+			npipeURL, err := url.Parse(npipePath)
+			if err != nil {
 				return nil, err
 			}
 
-			return net.Listen("unix", socketPath) // override tcp to use unix socket
+			return sshclient.ListenNpipe(npipeURL) // override tcp to use named pipe
 		}
 		p.AddRoute(local, &tcpproxy.DialProxy{
 			Addr:        remoteAddr,
diff --git a/pkg/sshclient/npipe_unsupported.go b/pkg/sshclient/npipe_unsupported.go
index 01d68e4..e120b00 100644
--- a/pkg/sshclient/npipe_unsupported.go
+++ b/pkg/sshclient/npipe_unsupported.go
@@ -9,6 +9,6 @@ import (
 	"net/url"
 )
 
-func listenNpipe(socketURI *url.URL) (net.Listener, error) {
+func ListenNpipe(socketURI *url.URL) (net.Listener, error) {
 	return nil, errors.New("named pipes are not supported by this platform")
 }
diff --git a/pkg/sshclient/npipe_windows.go b/pkg/sshclient/npipe_windows.go
index 3bf29ff..ac83170 100644
--- a/pkg/sshclient/npipe_windows.go
+++ b/pkg/sshclient/npipe_windows.go
@@ -16,7 +16,7 @@ import (
 // Allow built-in admins and system/kernel components
 const SddlDevObjSysAllAdmAll = "D:P(A;;GA;;;SY)(A;;GA;;;BA)"
 
-func listenNpipe(socketURI *url.URL) (net.Listener, error) {
+func ListenNpipe(socketURI *url.URL) (net.Listener, error) {
 	user, err := user.Current()
 	if err != nil {
 		return nil, err
diff --git a/pkg/sshclient/ssh_forwarder.go b/pkg/sshclient/ssh_forwarder.go
index 9070cd4..79de497 100644
--- a/pkg/sshclient/ssh_forwarder.go
+++ b/pkg/sshclient/ssh_forwarder.go
@@ -140,7 +140,7 @@ func setupProxy(ctx context.Context, socketURI *url.URL, dest *url.URL, identity
 			return &SSHForward{}, err
 		}
 	case "npipe":
-		listener, err = listenNpipe(socketURI)
+		listener, err = ListenNpipe(socketURI)
 		if err != nil {
 			return &SSHForward{}, err
 		}
diff --git a/pkg/types/handshake.go b/pkg/types/handshake.go
index 8aa855a..e9aa780 100644
--- a/pkg/types/handshake.go
+++ b/pkg/types/handshake.go
@@ -3,9 +3,10 @@ package types
 type TransportProtocol string
 
 const (
-	UDP  TransportProtocol = "udp"
-	TCP  TransportProtocol = "tcp"
-	UNIX TransportProtocol = "unix"
+	UDP   TransportProtocol = "udp"
+	TCP   TransportProtocol = "tcp"
+	UNIX  TransportProtocol = "unix"
+	NPIPE TransportProtocol = "npipe"
 )
 
 type ExposeRequest struct {
```

This builds, but breaks unix-unix and unix-tcp forwarding, replacing it with namedpipe-unix forwarding.
It's broken, but much simpler to see what are the main changes. Adding npipe-unix together while keeping unix-unix needs a bit of code reorganization first.

**Comment by praveenkumar** [comment]
@cfergeau With following change also it doesn't work on windows, getting wired error now. Will check the windows specific tests and see how it works with win-proxy side.
```
failed to expose port npipe:////./pipe/podman.sock -> ssh-tunnel://core@192.168.127.2:22/run/podman/podman.sock?key=C%3A%5CUsers%5Cprkumar%5C.crc%5Cmachines%5Ccrc%5Cid_ecdsa: address ssh-tunnel://core@192.168.127.2:22/run/podman/podman.sock?key=C%3A%5CUsers%5Cprkumar%5C.crc%5Cmachines%5Ccrc%5Cid_ecdsa: too many colons in address
```

**Comment by anjannath** [comment]
Resolved via #3002
